accepted/2023/feature-switch-attributes.md

```diff
@@ -0,0 +1,766 @@
+
+```csharp
+UseDynamicCode(); // warns
+
+if (RuntimeFeature.IsDynamicCodeSupported)
+  UseDynamicCode(); // OK, no warning
+
+if (RuntimeFeature.IsDynamicCodeCompiled)
+  UseDynamicCode(); // OK, no warning in ILCompiler (for now this warns in the analyzer)
+
+[RequiresDynamicCode("Uses dynamically generated code")]
+static void UseDynamicCode() { }
+```
+
+The ILLink Roslyn analyzer has built-in support for treating `IsDynamicCodeSupported` as a guard for `RequiresDynamicCodeAttribute`, but has no other built-in support.
+
+## Goals
+
+- Allow libraries to define their own feature guard properties
+
+  Libraries should be able to introduce their own properties that can act as guards for `RequiresDynamicCodeAttribute`, or for other features that might produce warnings in the analyzer
+
+- Define an attribute-based model for such feature guards
+
+- Take into account how this would interact with an attribute-based model for feature switches
+
+  We will explore what an attribute-based model for feature switches would look like to ensure that it interacts well with a model for feature guards. It's possible that we would design both in conjunction if they are naturally related.
+
+- Teach the ILLink Roslyn analyzer to treat `IsDynamicCodeCompiled` as a guard for `RequiresDynamicCodeAttribute` using the same model we define for third-party libraries
+
+### Non-goals
+
+- Support branch elimination in the analyzer for all feature switches
+
+  The most important use case for the analyzer is analyzing libraries. Libraries typically don't bake in constants for feature switches, so the analyzer needs to consider all branches. It should support feature guards for features that produce warnings, but doesn't need to consider feature settings passed in from the project file to treat some branches as dead code.
+
+- Teach the ILLink Roslyn Analyzer about the substitution XML
+
+  We don't want to teach the analyzer to read the substitution XML. The analyzer is the first interaction that users typically have with trimming and AOT warnings. This should not be burdened by the XML format. Even if we did teach the analyzer about the XML, it would not solve the problem; for example, the analyzer must not globally assume that `IsDynamicCodeSupported` is false as ILCompiler does.
+
+- Define a model with the full richness of the supported OS platform attributes
+
+  We will focus initially on a model where feature switches are booleans that return `true` if a feature is enabled. We aren't considering supporting version checks, or feature switches of the opposite polarity (where `true` means a feature is disabled/unsupported). We will consider what this might look like just enough to gain confidence that our model could be extended to support these cases in the future, but won't design this fully in the first iteration.
+
+## Feature guard attribute
```

Comment:
It would help if you could list all the features you'll want to guard today. The only examples that currently exist seem to be around dynamic code. That doesn't seem to be too compelling, especially because those already have attributes.

Also, do you imagine that users can extend the set of features? Or are the features being guarded a closed set that is defined by the .NET platform?